### ISSUE TITLE
Enumerating Telegram group chats; joining group chats with MixedCase And Space in their jid

### ISSUE THREAD

**Issue by ivucica** [SpectrumIM/spectrum2#299]
I just pulled the latest Docker image of Spectrum2, and I can't see `disco#items` advertised nor is there a useful response to sending it out. (See #190 claiming it should work.)

```

<iq xmlns="jabber:client" to="telegram.my.domain" type="get" id="8db74a9f-f667-48c8-b83d-3ccc458bc67f" from="localpart@my.domain/GajimWindows">
<query xmlns="http://jabber.org/protocol/disco#items" />
</iq>

<!-- Incoming 21/07/2018 20:21:44 -->
<iq id='8db74a9f-f667-48c8-b83d-3ccc458bc67f' type='result' to='localpart@my.domain/GajimWindows' from='telegram.my.domain'>
<query xmlns='http://jabber.org/protocol/disco#items'/>
</iq>
<r xmlns='urn:xmpp:sm:3'/>

<!-- Outgoing 21/07/2018 20:21:50 -->
<iq xmlns="jabber:client" to="telegram.my.domain" type="get" id="8db74a9f-f667-48c8-b83d-3ccc458bc67f" from="localpart@my.domain/GajimWindows">
<query xmlns="http://jabber.org/protocol/disco#info" />
</iq>

<!-- Incoming 21/07/2018 20:21:50 -->
<iq id='8db74a9f-f667-48c8-b83d-3ccc458bc67f' type='result' to='localpart@my.domain/GajimWindows' from='telegram.my.domain'>
<query xmlns='http://jabber.org/protocol/disco#info'>
<identity type='telegram' name='Spectrum Telegram Transport' category='gateway'/>
<feature var='jabber:iq:register'/>
<feature var='jabber:iq:gateway'/>
<feature var='jabber:iq:private'/>
<feature var='http://jabber.org/protocol/disco#info'/>
<feature var='http://jabber.org/protocol/commands'/>
<feature var='http://jabber.org/protocol/muc'/>
</query>
</iq>
```

Additionally, while clients that I tried (Gajim, Poezio) have allowed me to join `testgroupchat@telegram.my.domain`, I could not join `ABC@telegram.my.domain` or `Group Chat@telegram.my.domain`.

This seems to be by design; I checked, and it is:
- _expected_ that the room name will be lowercased, and
- it is also expected that whitespace will be escaped into `\20`

This is similar to #175 and #290.

- How do I enumerate all group chats on purple-telegram? If this isn't supported, can it be supported? (#190 says it should be.)
- How do I join a group chat named "Group Chat"? If this isn't supported, can it be supported? (This may depend on #290, #175.)

**Comment by ivucica** [comment]
Correction: This applies to purple-telegram. I get groupchats (although not pretty-named!) on purple-hangouts.

**Comment by ivucica** [comment]
To expand: I am actually also unable to send any messages to `purple-hangouts` groupchats. I'm presuming this is because the room names _also_ contain some uppercase symbols. So, groupchats are enumerated, but are not usable.

Docker image I use is spectrum2/spectrum@sha256:a2cad70f2a93f6d7d643bf02fd9b5a05f26a403727209e6611f4b8134425b278 from yesterday. I believe this to be built from Spectrum2 6d220b0a083b3adf7fd4ae65ca011b29488493b7.

I've tried installing purple-hangouts and purple-telegram on Windows using random installers I found published (no idea about revision they were built from). Group chat enumeration and joins works perfectly on Windows.

**Comment by vitalyster** [comment]
I’m unsure what spectrum can do with case sensitive room ids at all while xmpp jids are case insensitive, do you have any idea?

**Comment by vitalyster** [comment]
Anyway it’s duplicate of #175

**Comment by ivucica** [comment]
This is not a duplicate.

1. #175 does not mention lack of `disco#items` capability advertised for `purple-telegram`.
1. #175 does not mention problems with enumeration of chatrooms.
1. purple-hangouts is also affected, which #175 does not mention.

I disagree with closing this item, unless #175 is seriously expanded in scope. If you are up for that, happy to move the discussion there.

What's possible to do is make use of what's mentioned in #175: "spectrum have some support of mapping case-sensitive name to case-insensitive JID". 

- Upon connection, enumerate the chatrooms correctly when underlying plugins have support for it (or when underlying plugins send an update)
- Store these chatrooms in a map string->string (and reverse):
  - from lowercased, XEP-0106-escaped ID (used as localpart)
  - to the actual ID (used by the underlying plugin)
- How to resolve conflicts? Generate a UUID in case of a conflict and use it as localpart
- When an enumeration request comes, use the generated JIDs
- When joining, accept the generated JID and pass the actual ID to the underlying plugin
